### PR TITLE
persist: check if state is from the future when decoding it

### PR DESCRIPTION
**Pull request by danhhz**
If persist gets some encoded ProtoState from the future (e.g. two
versions of code are running simultaneously against the same shard), it
might have a field that the current code doesn't know about. This would
be silently discarded at proto decode time. Unknown Fields [1] are a
tool we can use in the future to help deal with this, but in the
short-term, it's best to keep the persist read-modify-CaS loop simple
for as long as we can get away with it (i.e. until we have to offer the
ability to do rollbacks).

[1] https://developers.google.com/protocol-buffers/docs/proto3#unknowns

To detect the bad situation and disallow it, we tag every version of
state written to consensus with the version of code used to encode it.
Then at decode time, we're able to compare the current version against
any we receive and assert as necessary. Initially we reject any version
from the future (most conservative but easiest to reason about). If/when
we support deploy rollbacks and rolling upgrades, we can adjust this
assert as necessary to reflect the policy.

We could do the same for blob data, but it shouldn't be necessary. Any
blob data we read is going to be because we fetched it using a pointer
stored in some persist state. If we can handle the state, we can handle
the blobs it references, too.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by danhhz** [comment]
cc @benesch you mentioned in our 1:1 that this seemed okay to do. is there anyone on cloud we should give a heads up to, given that this may cause the "old" storaged and computed processes to start panicing after environmentd is rolled in a version upgrade?

**Comment by benesch** [comment]
> cc @benesch you mentioned in our 1:1 that this seemed okay to do. is there anyone on cloud we should give a heads up to, given that this may cause the "old" storaged and computed processes to start panicing after environmentd is rolled in a version upgrade?

I wouldn't worry about it! There's other places where things can panic during a version upgrade today.

File: src/persist-client/src/impl/encoding.rs [hunk 1]
```diff
@@ -98,14 +99,46 @@ where
...
+        //
+        // [1]: https://developers.google.com/protocol-buffers/docs/proto3#unknowns
+        //
+        // To detect the bad situation and disallow it, we tag every version of
+        // state written to consensus with the version of code used to encode
+        // it. Then at decode time, we're able to compare the current version
+        // against any we receive and assert as necessary. Initially we reject
+        // any version from the future (most conservative but easiest to reason
+        // about). If/when we support deploy rollbacks and rolling upgrades, we
+        // can adjust this assert as necessary to reflect the policy.
```
Comment (pH14) [inline src/persist-client/src/impl/encoding.rs:128]: wouldn't mind highlighting the current policy + how it'll need to change in the future in a separate paragraph

**Review by pH14 — APPROVED** [review]
LGTM! huge fan of putting version numbers everywhere

File: src/persist-client/src/impl/encoding.rs (hunk 1 above)
Comment (danhhz) [reply src/persist-client/src/impl/encoding.rs]: done!

**Pull request merged**